src/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,433 @@
+        internal PreserveReferencesHandling PreserveHandlingOnDeserialize => _preserveHandlingOnDeserialize;
+        internal ReferenceLoopHandling LoopHandling => _loopHandling;
+
+        private PreserveReferencesHandling _preserveHandlingOnSerialize;
+        private PreserveReferencesHandling _preserveHandlingOnDeserialize;
+        private ReferenceLoopHandling _loopHandling;
+
+        public ReferenceHandling(PreserveReferencesHandling handling) : this(handling, handling, ReferenceLoopHandling.Error) { }
+
+        // For future, someone may want to define their own custom Handler with different behaviors of PreserveReferenceHandling on Serialize vs Deserialize and another ReferenceLoopHandling, such as ignore if added in a future.
+        private ReferenceHandling(PreserveReferencesHandling preserveHandlingOnSerialize, PreserveReferencesHandling preserveHandlingOnDeserialize, ReferenceLoopHandling loopHandling)
+        {
+            _preserveHandlingOnSerialize = preserveHandlingOnSerialize;
+            _preserveHandlingOnDeserialize = preserveHandlingOnDeserialize;
+            _loopHandling = loopHandling;
+        }
+    }
+
+    internal enum ReferenceLoopHandling
+    {
+        Error = 0,
+        // For future if requested by the community.
+        // Ignore = 1,
+    }
+
+    public enum PreserveReferencesHandling
+    {
+        None = 0,
+        All = 1,
+        // For future if requested by the community.
+        // Objects = 2,
+        // Arrays = 3,
+    }
+}
+
+namespace System.Text.Json
+{
+    public partial class JsonSerializerOptions
+    {
+        public ReferenceHandling ReferenceHandling { get; set; } = ReferenceHandling.Default;
+    }
+}
+```
+For System.Text.Json, the goal is to stick to the same *metadata* semantics for preserve from Json.Net and provide a similar usage in `JsonSerializerOptions` that encompasses at least the needed options for an immediate solution, which is to provide reference preservation.
+
+This API is about exposing the `ReferenceHandling` property as a class type, in order to make it very extensible for future; and provide built-in static instances of `Default` and `Preserve` that are useful to enable the most common behaviors by just setting those in `JsonSerializerOptions.ReferenceHandling`.
+
+With ReferenceHandling being a class, we can exclude things that, as of now, we are not sure if they need to be shipped, in example, the `ReferenceLoopHandling.Ignore` option or the Object and Array granularity of Json.Net's `PreserveReferencesHandling`.
+
+In a future, we might also be able to extend the Deserialization behaviors, in order to provide the granularity of Serialization on Deserialization. Or we could also be able to define independant behaviors for both Serialization and Deserialization; i.e: someone may want to emit payloads with preserved references but he does not want to read them.
+
```

Comment:
Also add a note that we considered enabling this feature via attributes (maybe with sample code that you had, if it worked), but deferred that to the future. If we can't do it for some reason, mention that.